samcli/lib/build/app_builder.py

```diff
@@ -269,6 +273,39 @@ def _get_build_graph(
         build_graph.clean_redundant_definitions_and_update(not self._is_building_specific_resource)
         return build_graph
 
+    def _validate_function_definition(self, function: Function) -> None:
+        """
+        Validates the properties of a function definition.
+
+        Raises an InvalidFunctionDefinition if the supplied function is invalid.
+
+        Parameters
+        ----------
+        function: Function
+            The named tuple representing the properties of the Lambda function definition to validate.
+
+        Raises
+        ------
+        InvalidFunctionDefinition
+            If the supplied function is invalid.
+
+        """
+        MIN_MEMORY_SIZE = 128
+        MAX_MEMORY_SIZE = 10240
+
+        if function.memory:
+            try:
+                function_memory = int(function.memory)
+            except ValueError as ex:
+                raise InvalidFunctionDefinition(
+                    f'Function {function.function_id} has invalid MemorySize "{function.memory}". Must be an integer between 128 and 10240.'
+                ) from ex
+
+            if function_memory < MIN_MEMORY_SIZE or function_memory > MAX_MEMORY_SIZE:
+                raise InvalidFunctionDefinition(
+                    f"Function {function.function_id} has MemorySize {function.memory}, but it must be an integer between 128 and 10240."
+                )
```

Comment:
Can we give a warning message rather than failing with the exception here? My concern is if this is changed and we missed it, then customers will be blocked building their lambda functions.

Reply:
Good point, however with Wing's comment we'll only be relying on `cfnlint` which we can expect to stay in sync with the CFN rules?